pandas/tests/frame/test_constructors.py

```diff
@@ -2457,6 +2457,33 @@ def test_constructor_list_str_na(self, string_dtype):
         expected = DataFrame({"A": ["1.0", "2.0", None]}, dtype=object)
         tm.assert_frame_equal(result, expected)
 
+    def test_constructor_large_size_frame(self):
+        # GH#50708
+        class LargeFrame(DataFrame):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+                for col in self.columns:
+                    if self.dtypes[col] == "O":
+                        self[col] = pd.to_numeric(self[col], errors="ignore")
+
+            @property
+            def _constructor(self):
+                return type(self)
+
+        def get_frame(N):
+            return LargeFrame(
+                data=np.vstack(
+                    [
+                        np.where(np.random.rand(N) > 0.36, np.random.rand(N), np.nan)
+                        for _ in range(2)
+                    ]
+                ).T
+            )
+
+        frame = get_frame(1000000)
+        # check that dropna() doesn't fall into an infinite loop
+        frame.dropna()
```

Comment:
Is there a way of breaking out of the infinite loop? This might cause weird ci Timeouts if we run into this but again in the future

Reply:
good point - in the notebook where I did git bisect I did
```python
class TimeoutError(Exception):
    pass

def timeout(seconds=10, error_message=os.strerror(errno.ETIME)):
    def decorator(func):
        def _handle_timeout(signum, frame):
            raise TimeoutError(error_message)

        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            signal.signal(signal.SIGALRM, _handle_timeout)
            signal.alarm(seconds)
            try:
                result = func(*args, **kwargs)
            finally:
                signal.alarm(0)
            return result

        return wrapper

    return decorator

@timeout(5)
def long_running_function(n):
    get_frame(n).dropna()
```

, taken from https://stackoverflow.com/a/2282656/4451315

Perhaps this test could be rewritten to use that, with a comment indicating why (and linking to the stackoverflow answer)

Reply:
Thank you, @MarcoGorelli. I will rewrite the test to avoid ci Timeout.

Reply:
If the previous code was failing with an recursion error, I would recommend setting `sys.setrecursionlimit` instead (I've done this with other tests that used to recursively fail

```
        rec_limit = sys.getrecursionlimit()
        try:
            sys.setrecursionlimit(100)
            code_that_fails()
        finally:
            sys.setrecursionlimit(rec_limit)
```

Reply:
I think this was just an infinite loop iirc, can check later

Reply:
@mroeschke any objections to the current solution?

When I tried `sys.setrecursionlimit` it didn't work for me

Reply:
I am still a little weary about adding a test like this despite the timeout protection. Is it known what cause the infinite loop before?

Reply:
sure, good point

From https://www.kaggle.com/code/marcogorelli/pandas-regression-example?scriptVersionId=117662725, I'm seeing https://github.com/pandas-dev/pandas/pull/49551 as the commit that fixed it, but I don't really see why that would be the case